[CANN] Fix the ACL_ERROR_REPEAT_INITIALIZE error that occurs when coexisting…

bachelor-dou opened the pull request:
### Description
This PR fixes the ACL_ERROR_REPEAT_INITIALIZE error that occurs when onnxruntime-cann coexists with torch_npu, which causes the CANN provider to fail to initialize and fall back to CPU for inference.



### Error log
```
2025-08-30 03:30:44.189484484 [E:onnxruntime:Default, provider_bridge_ort.cc:2279 TryGetProviderInfo_CANN] ~/code/onnxruntime/onnxruntime/core/providers/cann/cann_call.cc:143 bool onnxruntime::CannCall(ERRTYPE, const char*, const char*, ERRTYPE, const char*) [with ERRTYPE = int; bool THRW = true]
~/code/onnxruntime/onnxruntime/core/providers/cann/cann_call.cc:137 bool onnxruntime::CannCall(ERRTYPE, const char*, const char*, ERRTYPE, const char*) [with ERRTYPE = int; bool THRW = true] CANN failure 100002: ACL_ERROR_REPEAT_INITIALIZE ; NPU=0 ; hostname=coder-6a95445b-e353-4b0b-970b-41797b02ca23-566969c97-xlf2s ; expr=aclInit(nullptr); 


*************** EP Error ***************
EP Error /home/dou/code/onnxruntime/onnxruntime/python/onnxruntime_pybind_state.cc:1231 std::shared_ptr<onnxruntime::IExecutionProviderFactory> onnxruntime::python::CreateExecutionProviderFactoryInstance(const onnxruntime::SessionOptions&, const string&, const ProviderOptionsMap&) create CANN ExecutionProvider fail
 when using [('CANNExecutionProvider', {'device_id': 0, 'arena_extend_strategy': 'kNextPowerOfTwo', 'npu_mem_limit': 4294967296, 'enable_cann_graph': True}), 'CPUExecutionProvider']
Falling back to ['CPUExecutionProvider'] and retrying.
****************************************
================= before run
[array([[ -8.8664875,  -7.903085 ,  -6.529765 ,  -6.0811057,  -4.148506 ,
         -6.4790154,  -5.8431354,  -8.300879 ,  -6.77909  ,  -8.013498 ,
         -8.759175 ,  -7.237462 ,  -6.890937 ,  -8.367442 ,  -7.565303 ,
```

snnn commented on onnxruntime/core/providers/cann/cann_call.cc:
```diff
@@ -121,28 +122,32 @@ const char* CannErrString<ge::graphStatus>(ge::graphStatus e) {
+      try {
+        char hostname[HOST_NAME_MAX];
+        if (gethostname(hostname, HOST_NAME_MAX) != 0)
+          snprintf(hostname, HOST_NAME_MAX, "%s", "?");
```
Security bug... Check snprinf's document.

snnn commented:
The title and description of this PR are misleading. This PR is not about updating `build.sh` to use Python 3.12. It's about changes in the CANN execution provider.

A better title would be: "Fix repeated ACL initialization and simplify model file search in CANN EP".

A better description would be:
"This PR addresses two issues in the CANN execution provider:
1.  It prevents a crash when `aclInitialize` is called multiple times by detecting the `ACL_ERROR_REPEAT_INITIALIZE` error and skipping the call to `aclFinalize`.
2.  It simplifies the model file search by replacing the regex-based file matching with a simpler string search."

snnn reviewed (commented):
In `onnxruntime/core/providers/cann/cann_utils.cc`, the implementation of `MatchFile` has been changed from a regex-based search to a simple string search. The new implementation looks for a file that contains the given `file_name` as a substring and has a `.om` extension. Is this change intended and correct? It seems reasonable, but I want to double-check.

snnn commented:
Please run clang-format and fix the "[Lint / Python format (pull_request)](https://github.com/microsoft/onnxruntime/actions/runs/18122460251/job/51634333488?pr=26193) issue

bachelor-dou commented:
> Sorry for the delay. In the last release we added EP ABI support, so that hardware vendors can build EPs independent of ONNX Runtime version. We recommend every EP switching to that. See [microsoft/onnxruntime-inference-examples#527](https://github.com/microsoft/onnxruntime-inference-examples/pull/527) for an example.

Updated.  Thank you for your review.

snnn commented:
> In addition, is it possible to create a separate repo for each EP in the microsoft organization, called something like onnxruntime-cuda, onnxruntime-cann, and so on, which can bring several benefits to ONNX Runtime?

I prefer to host the repos in the independent hardware vendors (IHVs) Github org , so the hardware companies will own the IP asset.
